File: android/src/main/java/expo/modules/assemblyai/ExpoAssemblyAIModule.kt
package expo.modules.assemblyai

import android.Manifest
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.interfaces.permissions.Permissions

/**
 * Android native two-way audio bridge for `expo-assemblyai`.
 *
 * Mirrors [ExpoAssemblyAIModule.swift]: same function/event names so the JS
 * layer is platform-agnostic. The WebSocket lives in JS; this module only moves
 * PCM in and out of the hardware.
 */
class ExpoAssemblyAIModule : Module(), TwoWayAudioEngine.Listener {
  private val engine = TwoWayAudioEngine(this)

  override fun definition() = ModuleDefinition {
    Name("ExpoAssemblyAI")

    Events(
      "onMicrophoneData",
      "onInputVolume",
      "onOutputVolume",
      "onPlaybackFinished",
      "onError",
    )

    OnDestroy { engine.stop() }

    AsyncFunction("requestRecordingPermissionsAsync") { promise: Promise ->
      Permissions.askForPermissionsWithPermissionsManager(
        appContext.permissions, promise, Manifest.permission.RECORD_AUDIO,
      )
    }

    AsyncFunction("getRecordingPermissionsAsync") { promise: Promise ->
      Permissions.getPermissionsWithPermissionsManager(
        appContext.permissions, promise, Manifest.permission.RECORD_AUDIO,
      )
    }

    AsyncFunction("initialize") { options: AudioConfig ->
      engine.configure(
        TwoWayAudioEngine.Config(
          inputSampleRate = options.inputSampleRate,
          outputSampleRate = options.outputSampleRate,
          enableEchoCancellation = options.enableEchoCancellation,
          chunkDurationMs = options.chunkDurationMs,
        )
      )
    }

    AsyncFunction("start") {
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO)
        != PackageManager.PERMISSION_GRANTED
      ) {
        throw Exceptions.MissingPermissions(Manifest.permission.RECORD_AUDIO)
      }
      engine.start()
    }

    AsyncFunction("stop") { engine.stop() }
    Function("setMuted") { muted: Boolean -> engine.setMuted(muted) }
    Function("enqueuePlayback") { base64: String -> engine.enqueuePlayback(base64) }
    Function("clearPlayback") { engine.clearPlayback() }
  }

  override fun onCapture(base64Chunk: String) = sendEvent("onMicrophoneData", mapOf("data" to base64Chunk))
  override fun onInputLevel(level: Float) = sendEvent("onInputVolume", mapOf("level" to level))
  override fun onOutputLevel(level: Float) = sendEvent("onOutputVolume", mapOf("level" to level))
  override fun onPlaybackFinished() = sendEvent("onPlaybackFinished", emptyMap<String, Any>())
  override fun onError(message: String) = sendEvent("onError", mapOf("message" to message))
}

class AudioConfig : Record {
  @Field var inputSampleRate: Int = 16_000
  @Field var outputSampleRate: Int = 24_000
  @Field var enableEchoCancellation: Boolean = true
  @Field var chunkDurationMs: Int = 100
}

File: ios/ExpoAssemblyAIModule.swift
import AVFoundation
import ExpoModulesCore

/// Native two-way audio bridge for `expo-assemblyai`.
///
/// Owns a `TwoWayAudioEngine` (full-duplex mic + speaker with echo cancellation)
/// and exposes it to JS as functions + events. The JS `StreamingTranscriber` /
/// `VoiceAgent` clients own the WebSocket; this module only moves PCM in and out
/// of the hardware. Keeping the transport in JS (not native) is deliberate — it
/// keeps the native surface tiny, testable, and identical across iOS/Android.
public class ExpoAssemblyAIModule: Module, TwoWayAudioEngineDelegate {
  private let audio = TwoWayAudioEngine()

  public func definition() -> ModuleDefinition {
    Name("ExpoAssemblyAI")

    Events(
      "onMicrophoneData",  // { data: base64 PCM16 }
      "onInputVolume",  // { level: 0…1 }
      "onOutputVolume",  // { level: 0…1 }
      "onPlaybackFinished",  // {}
      "onError"  // { message }
    )

    OnCreate { audio.delegate = self }
    OnDestroy { audio.stop() }

    // MARK: Permissions

    AsyncFunction("requestRecordingPermissionsAsync") { (promise: Promise) in
      let resolve: (Bool) -> Void = { granted in
        promise.resolve(["granted": granted, "canAskAgain": true, "status": granted ? "granted" : "denied"])
      }
      // `AVAudioApplication` is iOS 17+; fall back to the deprecated
      // `AVAudioSession` permission API on older systems.
      if #available(iOS 17.0, *) {
        AVAudioApplication.requestRecordPermission(completionHandler: resolve)
      } else {
        AVAudioSession.sharedInstance().requestRecordPermission(resolve)
      }
    }

    Function("getRecordingPermissionsAsync") { () -> [String: Any] in
      // Normalize both the iOS 17+ (`AVAudioApplication`) and legacy
      // (`AVAudioSession`) record-permission enums to undetermined/granted/denied.
      let isGranted: Bool
      let isUndetermined: Bool
      if #available(iOS 17.0, *) {
        let status = AVAudioApplication.shared.recordPermission
        isGranted = status == .granted
        isUndetermined = status == .undetermined
      } else {
        let status = AVAudioSession.sharedInstance().recordPermission
        isGranted = status == .granted
        isUndetermined = status == .undetermined
      }
      return [
        "granted": isGranted,
        "canAskAgain": isUndetermined,
        "status": isGranted ? "granted" : (isUndetermined ? "undetermined" : "denied"),
      ]
    }

    // MARK: Session control

    AsyncFunction("initialize") { (options: AudioConfig) in
      audio.configure(
        .init(
          // Sample rates cross the JS bridge as integer Hz (matching the Android
          // module and the TS `AudioSessionConfig`); the engine wants Double for
          // AVAudioFormat, so widen here at the boundary.
          inputSampleRate: Double(options.inputSampleRate),
          outputSampleRate: Double(options.outputSampleRate),
          enableEchoCancellation: options.enableEchoCancellation,
          chunkDurationMs: options.chunkDurationMs))
    }

    AsyncFunction("start") { try audio.start() }
    AsyncFunction("stop") { audio.stop() }
    Function("setMuted") { (muted: Bool) in audio.setMuted(muted) }

    // MARK: Playback (agent audio)

    Function("enqueuePlayback") { (base64: String) in audio.enqueuePlayback(base64) }
    Function("clearPlayback") { audio.clearPlayback() }
  }

  // MARK: - TwoWayAudioEngineDelegate

  func audioEngine(_ engine: TwoWayAudioEngine, didCapture base64Chunk: String) {
    sendEvent("onMicrophoneData", ["data": base64Chunk])
  }
  func audioEngine(_ engine: TwoWayAudioEngine, didUpdateInputLevel level: Float) {
    sendEvent("onInputVolume", ["level": level])
  }
  func audioEngine(_ engine: TwoWayAudioEngine, didUpdateOutputLevel level: Float) {
    sendEvent("onOutputVolume", ["level": level])
  }
  func audioEngineDidFinishPlayback(_ engine: TwoWayAudioEngine) {
    sendEvent("onPlaybackFinished", [:])
  }
  func audioEngine(_ engine: TwoWayAudioEngine, didFail message: String) {
    sendEvent("onError", ["message": message])
  }
}

/// Mirrors the JS `AudioSessionConfig`. Field names/defaults must stay in sync
/// with `src/ExpoAssemblyAIModule.types.ts`.
struct AudioConfig: Record {
  @Field var inputSampleRate: Int = 16_000
  @Field var outputSampleRate: Int = 24_000
  @Field var enableEchoCancellation: Bool = true
  @Field var chunkDurationMs: Int = 100
}
